### PR TITLE
alternative to einsum which is slow in extreme cases

### PR DESCRIPTION
**Pull request by marcelbischoff**
I don't really like two code options and not liking the `g` loop. I anyways leave it here. It is 10% slower on `efficientnet` but 5x faster `test_net_speed.py`. Probably needing to write own tensorop anyways.

### CODE REVIEW

File: examples/efficientnet.py [hunk 1]
```diff
@@ -134,22 +134,22 @@ def load_weights_from_torch(self):
...
+  #import matplotlib.pyplot as plt
+  #plt.imshow(img[0].mean(axis=0))
+  #plt.show()
 
   # category labels
   import ast
   lbls = fetch("https://gist.githubusercontent.com/yrevar/942d3a0ac09ec9e5eb3a/raw/238f720ff059c1f82f368259d1ca4ffa5dd8f9f5/imagenet1000_clsidx_to_labels.txt")
   lbls = ast.literal_eval(lbls.decode('utf-8'))
 
+  print(lbls)
```
Comment (geohot) [inline examples/efficientnet.py:151]: Remove?

**Comment by geohot** [comment]
I don't want two code paths, how slow is always tensor dot?

File: examples/efficientnet.py (hunk 1 above)
Comment (marcelbischoff) [reply examples/efficientnet.py:151]: Yes.

**Comment by marcelbischoff** [comment]
> I don't want two code paths, how slow is always tensor dot?

As far as I can see, less than 10%

```testing [(1, 1, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.04 / 0.16 ms  bp: 0.16 / 2.42 ms
testing [(1, 1, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.05 / 0.17 ms  bp: 0.11 / 2.20 ms
testing [(1, 1, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.18 / 0.20 ms  bp: 0.30 / 1.77 ms
testing [(1, 1, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.11 / 0.16 ms  bp: 0.32 / 1.88 ms
testing [(1, 1, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.11 / 0.16 ms  bp: 0.34 / 1.68 ms
testing [(1, 3, 11, 28), (6, 3, 2, 2)]   torch/tinygrad fp: 0.06 / 0.19 ms  bp: 0.11 / 2.57 ms
testing [(1, 3, 11, 28), (6, 3, 2, 3)]   torch/tinygrad fp: 0.04 / 0.16 ms  bp: 0.14 / 2.64 ms
testing [(1, 3, 11, 28), (6, 3, 2, 5)]   torch/tinygrad fp: 0.06 / 0.17 ms  bp: 0.12 / 2.49 ms
testing [(1, 3, 11, 28), (6, 3, 5, 2)]   torch/tinygrad fp: 0.12 / 0.18 ms  bp: 0.34 / 1.92 ms
testing [(1, 3, 11, 28), (6, 3, 5, 3)]   torch/tinygrad fp: 0.15 / 0.18 ms  bp: 0.27 / 1.88 ms
testing [(1, 3, 11, 28), (6, 3, 5, 5)]   torch/tinygrad fp: 0.12 / 0.18 ms  bp: 0.51 / 1.81 ms
testing [(1, 3, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.12 / 0.24 ms  bp: 0.31 / 2.82 ms
testing [(1, 3, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.13 / 0.28 ms  bp: 0.30 / 2.65 ms
testing [(1, 3, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.11 / 0.25 ms  bp: 0.29 / 2.60 ms
testing [(1, 3, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.13 / 0.29 ms  bp: 0.30 / 1.98 ms
testing [(1, 3, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.12 / 0.31 ms  bp: 0.30 / 1.83 ms
testing [(1, 3, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.15 / 0.27 ms  bp: 0.30 / 1.78 ms
testing [(8, 1, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.15 / 0.26 ms  bp: 0.46 / 2.72 ms
testing [(8, 1, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.14 / 0.23 ms  bp: 0.69 / 2.86 ms
testing [(8, 1, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.15 / 0.26 ms  bp: 0.50 / 2.48 ms
testing [(8, 1, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.14 / 0.24 ms  bp: 0.46 / 2.32 ms
testing [(8, 1, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.14 / 0.24 ms  bp: 0.51 / 2.15 ms
testing [(8, 1, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.15 / 0.24 ms  bp: 0.49 / 2.21 ms
testing [(8, 3, 11, 28), (6, 3, 2, 2)]   torch/tinygrad fp: 0.15 / 0.30 ms  bp: 0.43 / 3.15 ms
testing [(8, 3, 11, 28), (6, 3, 2, 3)]   torch/tinygrad fp: 0.15 / 0.29 ms  bp: 0.58 / 3.09 ms
testing [(8, 3, 11, 28), (6, 3, 2, 5)]   torch/tinygrad fp: 0.16 / 0.30 ms  bp: 0.45 / 2.97 ms
testing [(8, 3, 11, 28), (6, 3, 5, 2)]   torch/tinygrad fp: 0.16 / 0.50 ms  bp: 0.48 / 2.53 ms
testing [(8, 3, 11, 28), (6, 3, 5, 3)]   torch/tinygrad fp: 0.16 / 0.35 ms  bp: 0.45 / 2.63 ms
testing [(8, 3, 11, 28), (6, 3, 5, 5)]   torch/tinygrad fp: 0.18 / 0.34 ms  bp: 0.69 / 2.65 ms
testing [(8, 3, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.15 / 0.40 ms  bp: 0.43 / 4.18 ms
testing [(8, 3, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.17 / 0.41 ms  bp: 0.46 / 4.15 ms
testing [(8, 3, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.19 / 0.38 ms  bp: 0.65 / 3.81 ms
testing [(8, 3, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.17 / 0.44 ms  bp: 0.43 / 3.22 ms
testing [(8, 3, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.22 / 0.47 ms  bp: 0.45 / 3.35 ms
testing [(8, 3, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.22 / 0.50 ms  bp: 0.56 / 3.21 ms

forward pass:  64.936 ms, 2.65x off baseline 24.477 ms
backward pass: 155.448 ms, 3.48x off baseline 44.636 ms

did inference in 0.76 s
```
einsum + ravel
```
testing [(1, 1, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.04 / 0.16 ms  bp: 0.16 / 2.42 ms
testing [(1, 1, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.05 / 0.17 ms  bp: 0.11 / 2.20 ms
testing [(1, 1, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.18 / 0.20 ms  bp: 0.30 / 1.77 ms
testing [(1, 1, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.11 / 0.16 ms  bp: 0.32 / 1.88 ms
testing [(1, 1, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.11 / 0.16 ms  bp: 0.34 / 1.68 ms
testing [(1, 3, 11, 28), (6, 3, 2, 2)]   torch/tinygrad fp: 0.06 / 0.19 ms  bp: 0.11 / 2.57 ms
testing [(1, 3, 11, 28), (6, 3, 2, 3)]   torch/tinygrad fp: 0.04 / 0.16 ms  bp: 0.14 / 2.64 ms
testing [(1, 3, 11, 28), (6, 3, 2, 5)]   torch/tinygrad fp: 0.06 / 0.17 ms  bp: 0.12 / 2.49 ms
testing [(1, 3, 11, 28), (6, 3, 5, 2)]   torch/tinygrad fp: 0.12 / 0.18 ms  bp: 0.34 / 1.92 ms
testing [(1, 3, 11, 28), (6, 3, 5, 3)]   torch/tinygrad fp: 0.15 / 0.18 ms  bp: 0.27 / 1.88 ms
testing [(1, 3, 11, 28), (6, 3, 5, 5)]   torch/tinygrad fp: 0.12 / 0.18 ms  bp: 0.51 / 1.81 ms
testing [(1, 3, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.12 / 0.24 ms  bp: 0.31 / 2.82 ms
testing [(1, 3, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.13 / 0.28 ms  bp: 0.30 / 2.65 ms
testing [(1, 3, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.11 / 0.25 ms  bp: 0.29 / 2.60 ms
testing [(1, 3, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.13 / 0.29 ms  bp: 0.30 / 1.98 ms
testing [(1, 3, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.12 / 0.31 ms  bp: 0.30 / 1.83 ms
testing [(1, 3, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.15 / 0.27 ms  bp: 0.30 / 1.78 ms
testing [(8, 1, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.15 / 0.26 ms  bp: 0.46 / 2.72 ms
testing [(8, 1, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.14 / 0.23 ms  bp: 0.69 / 2.86 ms
testing [(8, 1, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.15 / 0.26 ms  bp: 0.50 / 2.48 ms
testing [(8, 1, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.14 / 0.24 ms  bp: 0.46 / 2.32 ms
testing [(8, 1, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.14 / 0.24 ms  bp: 0.51 / 2.15 ms
testing [(8, 1, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.15 / 0.24 ms  bp: 0.49 / 2.21 ms
testing [(8, 3, 11, 28), (6, 3, 2, 2)]   torch/tinygrad fp: 0.15 / 0.30 ms  bp: 0.43 / 3.15 ms
testing [(8, 3, 11, 28), (6, 3, 2, 3)]   torch/tinygrad fp: 0.15 / 0.29 ms  bp: 0.58 / 3.09 ms
testing [(8, 3, 11, 28), (6, 3, 2, 5)]   torch/tinygrad fp: 0.16 / 0.30 ms  bp: 0.45 / 2.97 ms
testing [(8, 3, 11, 28), (6, 3, 5, 2)]   torch/tinygrad fp: 0.16 / 0.50 ms  bp: 0.48 / 2.53 ms
testing [(8, 3, 11, 28), (6, 3, 5, 3)]   torch/tinygrad fp: 0.16 / 0.35 ms  bp: 0.45 / 2.63 ms
testing [(8, 3, 11, 28), (6, 3, 5, 5)]   torch/tinygrad fp: 0.18 / 0.34 ms  bp: 0.69 / 2.65 ms
testing [(8, 3, 11, 28), (6, 1, 2, 2)]   torch/tinygrad fp: 0.15 / 0.40 ms  bp: 0.43 / 4.18 ms
testing [(8, 3, 11, 28), (6, 1, 2, 3)]   torch/tinygrad fp: 0.17 / 0.41 ms  bp: 0.46 / 4.15 ms
testing [(8, 3, 11, 28), (6, 1, 2, 5)]   torch/tinygrad fp: 0.19 / 0.38 ms  bp: 0.65 / 3.81 ms
testing [(8, 3, 11, 28), (6, 1, 5, 2)]   torch/tinygrad fp: 0.17 / 0.44 ms  bp: 0.43 / 3.22 ms
testing [(8, 3, 11, 28), (6, 1, 5, 3)]   torch/tinygrad fp: 0.22 / 0.47 ms  bp: 0.45 / 3.35 ms
testing [(8, 3, 11, 28), (6, 1, 5, 5)]   torch/tinygrad fp: 0.22 / 0.50 ms  bp: 0.56 / 3.21 ms

forward pass:  407.538 ms, 16.77x off baseline 24.307 ms
backward pass: 143.299 ms, 3.38x off baseline 42.434 ms

did inference in 0.70 s
```

**Pull request merged**